tka/tailchonk.go

```diff
@@ -156,3 +165,255 @@ updateLoop:
+	c.l.RLock()
+	defer c.l.RUnlock()
+
+	info, err := c.get(prevAUMHash)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// not knowing about this hash is not an error
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	out := make([]AUM, len(info.Children))
+	for i, h := range info.Children {
+		c, err := c.get(h)
+		if err != nil {
+			// We expect any AUM recorded as a child on its parent to exist.
+			return nil, fmt.Errorf("reading child %d of %x: %v", i, h, err)
+		}
+		if c.AUM == nil {
+			return nil, fmt.Errorf("child %d of %x: AUM not stored", i, h)
+		}
+		out[i] = *c.AUM
+	}
+
+	return out, nil
+}
+
+func (c *FS) scanHashes(eachHashInfo func(*fsHashInfo)) error {
+	prefixDirs, err := os.ReadDir(c.base)
+	if err != nil {
+		return fmt.Errorf("reading prefix dirs: %v", err)
+	}
+	for _, prefix := range prefixDirs {
+		if !prefix.IsDir() {
+			continue
+		}
+		pb, err := hex.DecodeString(prefix.Name())
+		if err != nil || len(pb) != 1 {
+			return fmt.Errorf("invalid prefix directory %q: %v", prefix.Name(), err)
+		}
+
+		files, err := os.ReadDir(filepath.Join(c.base, prefix.Name()))
+		if err != nil {
+			return fmt.Errorf("reading prefix dir: %v", err)
+		}
+		for _, file := range files {
+			remainingHash, err := base32.StdEncoding.WithPadding(base32.NoPadding).DecodeString(file.Name())
+			if err != nil {
+				return fmt.Errorf("invalid aum file %s/%s: %v", prefix.Name(), file.Name(), err)
+			}
+			var h AUMHash
+			h[0] = pb[0]
+			copy(h[1:], remainingHash)
+
+			info, err := c.get(h)
+			if err != nil {
+				return fmt.Errorf("reading %x: %v", h, err)
+			}
+
+			eachHashInfo(info)
+		}
+	}
+
+	return nil
+}
+
+// Heads returns AUMs for which there are no children. In other
+// words, the latest AUM in all possible chains (the 'leaves').
+//
+// Heads is expected to be called infrequently compared to AUM() or
+// ChildAUMs(), so we haven't put any work into maintaining an index.
+// Instead, the full set of AUMs is scanned.
+func (c *FS) Heads() ([]AUM, error) {
+	c.l.RLock()
+	defer c.l.RUnlock()
+
+	out := make([]AUM, 0, 6)
+	err := c.scanHashes(func(info *fsHashInfo) {
+		if len(info.Children) == 0 && info.AUM != nil {
+			out = append(out, *info.AUM)
+		}
+	})
+	return out, err
+}
+
+// SetLastActiveAncestor is called to record the oldest-known AUM
+// that contributed to the current state. This value is used as
+// a hint on next startup to determine which chain to pick when computing
+// the current state, if there are multiple distinct chains.
+func (c *FS) SetLastActiveAncestor(hash AUMHash) error {
+	c.l.Lock()
+	defer c.l.Unlock()
+	return atomicfile.WriteFile(filepath.Join(c.base, "last_active_ancestor"), hash[:], 0644)
+}
+
+// LastActiveAncestor returns the oldest-known AUM that was (in a
+// previous run) an ancestor of the current state. This is used
+// as a hint to pick the correct chain in the event that the Chonk stores
+// multiple distinct chains.
+func (c *FS) LastActiveAncestor() (*AUMHash, error) {
```

Comment:
I'm using the pointer type as a stand-in for rusts' `Option<>` - Its possible for there to be no LastActiveAncestor set, which would return `(nil, nil)`.

The rest of the API uses the non-pointer `AUMHash` because it doesn't make sense for you to pass a nil hash.